src/FSharp.Core/local.fs

```diff
@@ -480,17 +480,57 @@ module internal List =
             else
                 filterToFreshConsTail cons f t
 
-    let rec filter predicate l =
-        match l with
-        | [] -> l
-        | h :: ([] as nil) -> if predicate h then l else nil
-        | h :: t ->
-            if predicate h then
-                let cons = freshConsNoTail h
-                filterToFreshConsTail cons predicate t
-                cons
-            else
-                filter predicate t
+    [<System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.NoInlining)>]
+    let invalidStateOfEmptyList () = raise (System.Exception "logic error; list should not be empty")
+
+    /// optimized filter if filtered items are all at the end of the list then that list is returned
```

Comment:
I must say I find this a little more difficult to reason about, but I fully admit this is a "me" issue.
My natural tendency to reason about this optimization would be to go backwards from the end of the list, keeping the original tail reference while all items are ` predicate item = true` when going backwards (mentally to say, using the recursive step to have this backwards processing).

So no matter how many segments of "true" or "false" items I head in the front, I only care about all the "true" items at the very end.

So in a naive (naive esp. because of missing tailcall optimization) way like this.
```
let rec myNaiveFilter inputList predicate =
    match inputList  with
    | [] -> inputList 
    | head :: [] when predicate head -> inputList 
    | _ :: [] -> []
    | head :: tail  ->
        let filteredTail = myNaiveFilter tail predicate
        match predicate head, obj.ReferenceEquals(tail,filteredTail) with
        | true,true -> inputList 
        | true, false -> head :: filteredTail
        | false, _ -> filteredTail
```

Is the proposed implementation doing effectively the same?
Or is the proposed a different optimization and the purpose is different?

Especially the call on line 512 `let filteredTail = entryToFilterToFreshConsTail remaining` somehow indicates me that this path still behaves as before, with new allocations in that case.